src/main.rs

```diff
@@ -1163,6 +1163,80 @@ mod tests {
         });
     }
 
+    #[test]
+    fn test_vhost_user_net_guest_host() {
+        test_block!(tb, "", {
+            let mut bionic = BionicDiskConfig::new();
+            let guest = Guest::new(&mut bionic);
+
+            let mut workload_path = dirs::home_dir().unwrap();
+            workload_path.push("workloads");
+
+            let mut vubridge_path = workload_path.clone();
+            vubridge_path.push("vubridge");
+            let vubridge_path = String::from(vubridge_path.to_str().unwrap());
+
+            // Start the daemon
+            let mut daemon_child = Command::new(vubridge_path.as_str()).spawn().unwrap();
+
+            let mut cloud_child = Command::new("target/debug/cloud-hypervisor")
+                .args(&["--cpus", "4"])
+                .args(&["--memory", "size=512M,file=/dev/shm"])
+                .args(&["--kernel", guest.fw_path.as_str()])
+                .args(&[
+                    "--disk",
+                    guest
+                        .disk_config
+                        .disk(DiskType::OperatingSystem)
+                        .unwrap()
+                        .as_str(),
+                    guest
+                        .disk_config
+                        .disk(DiskType::CloudInit)
+                        .unwrap()
+                        .as_str(),
+                ])
+                .args(&[
+                    "--vhost-user-net",
+                    "mac=52:54:00:02:d9:01,sock=/tmp/vubr.sock",
+                ])
+                .args(&["--net", guest.default_net_string().as_str()])
+                .spawn()
+                .unwrap();
+
+            guest.ssh_command("sudo ip addr add 192.168.222.1/24 dev enp0s4")?;
+            guest.ssh_command("nc -u -l 192.168.222.1 4444 > net.log &")?;
+
+            let data = "\\x52\\x54\\x00\\x02\\xd9\\x01\\xec\\xb1\\xd7\\x98\\x3a\\xc0\\x08\\x00\\x45\\x00\\x00\\x40\\x00\\x00\\x00\\x00\\x40\\x11\\x3d\\x58\\xc0\\xa8\\xde\\x02\\xc0\\xa8\\xde\\x01\\x11\\x5c\\x11\\x5c\\x00\\x2c\\xa0\\x5e\\x48\\x65\\x6c\\x6c\\x6f\\x20\\x57\\x6f\\x72\\x6c\\x64\\x20\\x74\\x68\\x72\\x6f\\x75\\x67\\x68\\x20\\x76\\x68\\x6f\\x73\\x74\\x2d\\x75\\x73\\x65\\x72\\x2d\\x6e\\x65\\x74\\x21\\x0a";
+            std::process::Command::new("echo")
```

Comment:
I don't get it. If you use nc through both sides of the connection the message should be preserved so you can just do:

`data = "Hello World through vhost-user-net!"`

Reply:
The reason is you simply cannot do that. On one hand, you need the packet to be delivered to the port 4444 the vubridge process is listening onto from the host, which is why you need to do nc -u localhost 4444, but on the other hand you want the frame to be tagged with the destination IP 192.168.222.1, which is the IP of the network interface inside the VM.